bus-mapping/src/evm/opcodes/calldataload.rs

```diff
@@ -0,0 +1,207 @@
+                }
+            })
+            .collect::<Vec<u8>>();
+
+        state.push_stack_op(
+            &mut exec_step,
+            RW::WRITE,
+            geth_step.stack.last_filled(),
+            U256::from_big_endian(&calldata_word),
+        )?;
+
+        Ok(vec![exec_step])
+    }
+}
+
+#[cfg(test)]
+mod calldataload_tests {
+    use eth_types::{
+        bytecode,
+        evm_types::{OpcodeId, StackAddress},
+        geth_types::GethData,
+        Word,
+    };
+    use mock::{
+        test_ctx::helpers::{account_0_code_account_1_no_code, tx_from_1_to_0},
+        TestContext,
+    };
+    use rand::random;
+
+    use crate::{circuit_input_builder::ExecState, mock::BlockData, operation::StackOp};
+
+    use super::*;
+
+    fn rand_bytes(size: usize) -> Vec<u8> {
+        (0..size).map(|_| random()).collect::<Vec<u8>>()
+    }
+
+    #[test]
+    fn calldataload_opcode_impl() {
+        // 1. should be right padded
+        test_ok(0u64, vec![1u8, 2u8], {
+            let mut v = vec![0u8; 32];
+            v[0] = 1u8;
+            v[1] = 2u8;
+            Word::from_big_endian(&v)
+        });
+        // 2. exactly 32 bytes
+        let calldata = rand_bytes(32);
+        test_ok(0u64, calldata.clone(), Word::from_big_endian(&calldata));
+        // 3. take only 32 bytes
+        let calldata = rand_bytes(64);
+        test_ok(
+            12u64,
+            calldata.clone(),
+            Word::from_big_endian(&calldata[12..44]),
+        );
+    }
+
+    fn test_ok(offset: u64, calldata: Vec<u8>, calldata_word: Word) {
+        let code = bytecode! {
+            PUSH32(offset)
+            CALLDATALOAD
+            STOP
+        };
+
+        let mut block: GethData = TestContext::<2, 1>::new(
+            None,
+            account_0_code_account_1_no_code(code),
+            tx_from_1_to_0,
+            |block, _tx| block.number(0xcafeu64),
+        )
+        .unwrap()
+        .into();
+        block.eth_block.transactions[0].input = calldata.into();
+
+        let mut builder = BlockData::new_from_geth_data(block.clone()).new_circuit_input_builder();
+        builder
+            .handle_block(&block.eth_block, &block.geth_traces)
+            .unwrap();
+
+        let step = builder.block.txs()[0]
+            .steps()
+            .iter()
+            .find(|step| step.exec_state == ExecState::Op(OpcodeId::CALLDATALOAD))
+            .unwrap();
+
+        assert_eq!(
+            [0, 2]
+                .map(|idx| &builder.block.container.stack[step.bus_mapping_instance[idx].as_usize()])
+                .map(|op| (op.rw(), op.op())),
+            [
+                (
+                    RW::READ,
+                    &StackOp::new(1, StackAddress::from(1023), Word::from(offset)),
+                ),
+                (
+                    RW::WRITE,
+                    &StackOp::new(1, StackAddress::from(1023), calldata_word),
+                ),
+            ]
+        );
+
+        assert_eq!(
```

Comment:
I'm missing testcases where we aren't in a root_call and so we can test the missing CallContext read ops for `CallDataOffset` and `CallerId`.

If you need some inspiration on how to perform Calls which trigger CALLDATALOAD, please check: https://github.com/appliedzkp/zkevm-circuits/blob/a9bdc9a87f9fcb8068fd4a3f254c94194a0f4e1d/bus-mapping/src/circuit_input_builder.rs#L2679-#L2702